File: src/lograg/core/log_processor.py
"""
Log processing utilities for cleaning, parsing, and preparing log data
"""
import re
import hashlib
from datetime import datetime
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
from enum import Enum


class LogLevel(Enum):
    """Log level enumeration"""
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARN = "WARN" 
    WARNING = "WARNING"
    ERROR = "ERROR"
    FATAL = "FATAL"
    CRITICAL = "CRITICAL"


@dataclass
class LogEntry:
    """Structured log entry"""
    content: str
    timestamp: Optional[datetime] = None
    level: Optional[LogLevel] = None
    source: Optional[str] = None
    metadata: Optional[Dict] = None
    
    @property
    def id(self) -> str:
        """Generate unique ID for the log entry"""
        content_hash = hashlib.md5(self.content.encode()).hexdigest()
        timestamp_str = self.timestamp.isoformat() if self.timestamp else ""
        return f"{timestamp_str}_{content_hash[:8]}"


class LogProcessor:
    """Process and clean log entries for embedding"""
    
    # Common log patterns
    TIMESTAMP_PATTERNS = [
        r'\d{4}-\d{2}-\d{2}[\sT]\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?',
        r'\d{2}/\d{2}/\d{4}\s\d{2}:\d{2}:\d{2}',
        r'\w{3}\s+\d{1,2}\s+\d{2}:\d{2}:\d{2}',
    ]
    
    LOG_LEVEL_PATTERN = r'\b(DEBUG|INFO|WARN(?:ING)?|ERROR|FATAL|CRITICAL)\b'
    
    # Common noise patterns to remove/normalize
    NOISE_PATTERNS = [
        (r'\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b', '[IP]'),  # IP addresses
        (r'\b[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}\b', '[UUID]'),  # UUIDs
        (r'\b\d{13,}\b', '[TIMESTAMP]'),  # Unix timestamps
        (r'\b[A-Za-z0-9+/=]{20,}\b', '[TOKEN]'),  # Tokens/keys
    ]
    
    def __init__(self, max_length: int = 1000):
        self.max_length = max_length
        
    def parse_log_entry(self, raw_log: str, source: Optional[str] = None) -> LogEntry:
        """Parse raw log string into structured LogEntry"""
        
        # Extract timestamp
        timestamp = self._extract_timestamp(raw_log)
        
        # Extract log level
        level = self._extract_log_level(raw_log)
        
        # Clean and normalize content
        content = self._clean_content(raw_log)
        
        return LogEntry(
            content=content,
            timestamp=timestamp,
            level=level,
            source=source
        )
    
    def _extract_timestamp(self, log_text: str) -> Optional[datetime]:
        """Extract timestamp from log text"""
        for pattern in self.TIMESTAMP_PATTERNS:
            match = re.search(pattern, log_text)
            if match:
                timestamp_str = match.group()
                try:
                    # Try different datetime formats
                    for fmt in [
                        '%Y-%m-%d %H:%M:%S.%f',
                        '%Y-%m-%d %H:%M:%S', 
                        '%Y-%m-%dT%H:%M:%S.%fZ',
                        '%Y-%m-%dT%H:%M:%SZ',
                        '%m/%d/%Y %H:%M:%S',
                        '%b %d %H:%M:%S'
                    ]:
                        try:
                            return datetime.strptime(timestamp_str, fmt)
                        except ValueError:
                            continue
                except Exception:
                    pass
        return None
    
    def _extract_log_level(self, log_text: str) -> Optional[LogLevel]:
        """Extract log level from log text"""
        match = re.search(self.LOG_LEVEL_PATTERN, log_text, re.IGNORECASE)
        if match:
            level_str = match.group().upper()
            if level_str == "WARN":
                level_str = "WARNING"
            try:
                return LogLevel(level_str)
            except ValueError:
                pass
        return None
    
    def _clean_content(self, log_text: str) -> str:
        """Clean and normalize log content"""
        content = log_text.strip()
        
        # Remove/normalize noise patterns
        for pattern, replacement in self.NOISE_PATTERNS:
            content = re.sub(pattern, replacement, content, flags=re.IGNORECASE)
        
        # Remove excessive whitespace
        content = re.sub(r'\s+', ' ', content)
        
        # Truncate if too long
        if len(content) > self.max_length:
            content = content[:self.max_length] + "..."
            
        return content
    
    def extract_key_terms(self, log_entry: LogEntry) -> List[str]:
        """Extract key terms from log entry for better indexing"""
        content = log_entry.content.lower()
        
        # Extract technical terms, error codes, etc.
        terms = []
        
        # Error codes (e.g., HTTP status codes)
        error_codes = re.findall(r'\b[45]\d{2}\b', content)
        terms.extend([f"status_{code}" for code in error_codes])
        
        # Exception names
        exceptions = re.findall(r'\b\w*(?:Exception|Error)\b', content, re.IGNORECASE)
        terms.extend(exceptions)
        
        # Service/component names (usually capitalized words)
        components = re.findall(r'\b[A-Z][a-zA-Z]+(?:Service|Manager|Handler|Controller)\b', log_entry.content)
        terms.extend(components)
        
        return list(set(terms))
    
    def chunk_long_log(self, log_entry: LogEntry, chunk_size: int = 500, overlap: int = 50) -> List[LogEntry]:
        """Split long log entries into chunks"""
        if len(log_entry.content) <= chunk_size:
            return [log_entry]
        
        chunks = []
        content = log_entry.content
        start = 0
        
        while start < len(content):
            end = start + chunk_size
            if end >= len(content):
                chunk_content = content[start:]
            else:
                # Try to break at word boundary
                space_idx = content.rfind(' ', start, end)
                if space_idx > start:
                    end = space_idx
                chunk_content = content[start:end]
            
            chunk_entry = LogEntry(
                content=chunk_content,
                timestamp=log_entry.timestamp,
                level=log_entry.level,
                source=log_entry.source,
                metadata={**(log_entry.metadata or {}), "chunk_start": start}
            )
            chunks.append(chunk_entry)
            
            start = end - overlap if end < len(content) else len(content)
            
        return chunks
    
    def batch_process(self, raw_logs: List[str], source: Optional[str] = None) -> List[LogEntry]:
        """Process multiple raw log entries"""
        processed = []
        for raw_log in raw_logs:
            try:
                entry = self.parse_log_entry(raw_log, source)
                processed.append(entry)
            except Exception as e:
                # Log processing error, but continue with others
                print(f"Error processing log: {e}")
                continue
        return processed

File: src/logProcessor.ts
/**
 * Log processing utilities
 */
import { createHash } from 'crypto';
import { LogEntry, LogLevel } from './types.js';

export class LogProcessor {
    private maxLength: number;

    // Common log patterns
    private static TIMESTAMP_PATTERNS = [
        /\d{4}-\d{2}-\d{2}[\sT]\d{2}:\d{2}:\d{2}(?:\.\d+)?(?:Z|[+-]\d{2}:\d{2})?/,
        /\d{2}\/\d{2}\/\d{4}\s\d{2}:\d{2}:\d{2}/,
        /\w{3}\s+\d{1,2}\s+\d{2}:\d{2}:\d{2}/,
    ];

    private static LOG_LEVEL_PATTERN = /\b(DEBUG|INFO|WARN(?:ING)?|ERROR|FATAL|CRITICAL)\b/i;

    // Noise patterns to normalize
    private static NOISE_PATTERNS: [RegExp, string][] = [
        [/\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b/g, '[IP]'], // IP addresses
        [/\b[a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12}\b/gi, '[UUID]'], // UUIDs
        [/\b\d{13,}\b/g, '[TIMESTAMP]'], // Unix timestamps
        [/\b[A-Za-z0-9+/=]{20,}\b/g, '[TOKEN]'], // Tokens/keys
    ];

    constructor(maxLength: number = 1000) {
        this.maxLength = maxLength;
    }

    /**
     * Parse raw log string into structured LogEntry
     */
    parseLogEntry(rawLog: string, source?: string): LogEntry {
        const timestamp = this.extractTimestamp(rawLog);
        const level = this.extractLogLevel(rawLog);
        const content = this.cleanContent(rawLog);
        const id = this.generateId(content, timestamp);

        return {
            id,
            content,
            ...(timestamp && { timestamp }),
            ...(level && { level }),
            ...(source && { source }),
        };
    }

    /**
     * Extract timestamp from log text
     */
    private extractTimestamp(logText: string): Date | undefined {
        for (const pattern of LogProcessor.TIMESTAMP_PATTERNS) {
            const match = logText.match(pattern);
            if (match) {
                try {
                    return new Date(match[0]);
                } catch (error) {
                    // Try parsing with different formats
                    continue;
                }
            }
        }
        return undefined;
    }

    /**
     * Extract log level from log text
     */
    private extractLogLevel(logText: string): LogLevel | undefined {
        const match = logText.match(LogProcessor.LOG_LEVEL_PATTERN);
        if (match) {
            const levelStr = match[1].toUpperCase();
            if (levelStr === 'WARN') return LogLevel.WARNING;
            return Object.values(LogLevel).find(level => level === levelStr);
        }
        return undefined;
    }

    /**
     * Clean and normalize log content
     */
    private cleanContent(logText: string): string {
        let content = logText.trim();

        // Apply noise patterns
        for (const [pattern, replacement] of LogProcessor.NOISE_PATTERNS) {
            content = content.replace(pattern, replacement);
        }

        // Remove excessive whitespace
        content = content.replace(/\s+/g, ' ');

        // Truncate if too long
        if (content.length > this.maxLength) {
            content = content.substring(0, this.maxLength) + '...';
        }

        return content;
    }

    /**
     * Generate unique ID for log entry
     */
    private generateId(content: string, timestamp?: Date): string {
        const hash = createHash('md5').update(content).digest('hex').substring(0, 8);
        const timestampStr = timestamp?.toISOString() || '';
        return `${timestampStr}_${hash}`;
    }

    /**
     * Extract key terms from log entry for better indexing
     */
    extractKeyTerms(entry: LogEntry): string[] {
        const content = entry.content.toLowerCase();
        const terms: string[] = [];

        // Error codes (HTTP status codes)
        const errorCodes = content.match(/\b[45]\d{2}\b/g) || [];
        terms.push(...errorCodes.map(code => `status_${code}`));

        // Exception names
        const exceptions = content.match(/\b\w*(?:exception|error)\b/gi) || [];
        terms.push(...exceptions);

        // Service/component names
        const components = entry.content.match(/\b[A-Z][a-zA-Z]+(?:Service|Manager|Handler|Controller)\b/g) || [];
        terms.push(...components);

        return [...new Set(terms)];
    }

    /**
     * Split long log entries into chunks
     */
    chunkLongLog(entry: LogEntry, chunkSize: number = 500, overlap: number = 50): LogEntry[] {
        if (entry.content.length <= chunkSize) {
            return [entry];
        }

        const chunks: LogEntry[] = [];
        let start = 0;

        while (start < entry.content.length) {
            let end = start + chunkSize;

            if (end >= entry.content.length) {
                end = entry.content.length;
            } else {
                // Try to break at word boundary
                const spaceIndex = entry.content.lastIndexOf(' ', end);
                if (spaceIndex > start) {
                    end = spaceIndex;
                }
            }

            const chunkContent = entry.content.substring(start, end);
            const chunkEntry: LogEntry = {
                ...entry,
                id: `${entry.id}_chunk_${start}`,
                content: chunkContent,
                metadata: {
                    ...entry.metadata,
                    chunkStart: start,
                    isChunk: true,
                    originalId: entry.id,
                },
            };

            chunks.push(chunkEntry);
            start = end - overlap;
            if (start >= entry.content.length) break;
        }

        return chunks;
    }

    /**
     * Process multiple raw log entries
     */
    batchProcess(rawLogs: string[], source?: string): LogEntry[] {
        const processed: LogEntry[] = [];

        for (const rawLog of rawLogs) {
            try {
                const entry = this.parseLogEntry(rawLog, source);
                processed.push(entry);
            } catch (error) {
                console.warn('Error processing log:', error);
                continue;
            }
        }

        return processed;
    }
}
